frontend/src/api/APIClient.ts

```diff
@@ -0,0 +1,119 @@
+/* eslint-disable class-methods-use-this */
+/* eslint-disable @typescript-eslint/no-explicit-any */
+import ApiError from './ApiError';
+
+type Method = 'GET' | 'POST' | 'PUT' | 'DELETE' | 'PATCH';
+
+type BodyHashMap = Record<string, any>;
+
+interface BaseAPIClientParams {
+  path: string;
+  hasCookies?: boolean;
+}
+
+interface APIClientParamsWithBody extends BaseAPIClientParams {
+  body?: BodyHashMap;
+}
+
+interface APIClientType {
+  get<T>(params: BaseAPIClientParams): Promise<T>;
+  post<T>(params: APIClientParamsWithBody): Promise<T>;
+  patch<T>(params: APIClientParamsWithBody): Promise<T>;
+  put<T>(params: APIClientParamsWithBody): Promise<T>;
+  delete(params: BaseAPIClientParams): Promise<void>;
+}
+
+export default class APIClient implements APIClientType {
+  private baseURL: URL;
+
+  constructor(baseURL: string) {
+    this.baseURL = new URL(baseURL);
+  }
+
+  async get<T>(params: BaseAPIClientParams): Promise<T> {
+    return this.request<T>({ method: 'GET', ...params });
+  }
+
+  async post<T>(params: APIClientParamsWithBody): Promise<T> {
+    return this.request<T>({ method: 'POST', ...params });
+  }
+
+  async patch<T>(params: APIClientParamsWithBody): Promise<T> {
+    return this.request<T>({ method: 'PATCH', ...params });
+  }
+
+  async put<T>(params: APIClientParamsWithBody): Promise<T> {
+    return this.request<T>({ method: 'PUT', ...params });
+  }
+
+  async delete(params: BaseAPIClientParams): Promise<void> {
+    return this.request<void>({ method: 'DELETE', ...params });
+  }
+
+  private getRequestInit({
+    method,
+    body,
+    hasCookies = true,
+  }: {
+    method: Method;
+    body?: BodyHashMap;
+    hasCookies?: boolean;
+  }) {
+    const headers: HeadersInit = {
+      Accept: 'application/json',
+    };
+
+    const requestInit: RequestInit = {
+      method,
+      credentials: hasCookies ? 'include' : 'omit',
+      headers,
+    };
+
+    if (['POST', 'PUT', 'PATCH'].includes(method)) {
+      headers['Content-Type'] = 'application/json';
+    }
+
+    if (body) {
+      requestInit.body = JSON.stringify(body);
+    }
+
+    return requestInit;
+  }
+
+  private async request<T>({
+    path,
+    method,
+    body,
+    hasCookies,
+  }: {
+    path: string;
+    method: Method;
+    body?: BodyHashMap;
+    hasCookies?: boolean;
+  }): Promise<T> {
+    const url = this.baseURL + path;
+    const response = await fetch(url, this.getRequestInit({ method, body, hasCookies }));
+
+    if (!response.ok) {
+      const { status, statusText } = response;
+      let errorMessage = `API통신에 실패했습니다: ${statusText}`;
+      // response가 JSON형식으로 응답을 주지 않을 경우를 대비하여 try - catch를 사용합니다.
+      try {
+        const errorData = await response.json();
+        errorMessage += ` - ${errorData.message}`;
+      } catch {
+        // JSON 파싱이 실패하면 기본 메시지만 사용합니다.
+      }
+      throw new ApiError({ message: errorMessage, statusCode: status, method });
```

Comment:
이 부분 catch가 하는 일이 없네요! 
response.json()이 실패했을 경우에 대비해서 try-catch 안에 넣어주신 것 같은데, 사용하지 않는 3 줄을 더 적는 것 보다 이렇게 작성하면 어떨지 제안드립니다. 
단순 제안이니 반영은 선택적으로 해주세요~

```suggestion
      const { status, statusText } = response;
      const defaultErrorMessage = `API통신에 실패했습니다: ${statusText}`;
  
      const errorData = await response.json().catch(() => null);
      const errorMessage = `${defaultErrorMessage}${errorData?.message ? ` - ${errorData.message}` : ''}`;
      
      throw new ApiError({ message: errorMessage, statusCode: status, method });
```